### PR TITLE
Feat/revamp rma request return

### PR DESCRIPTION
**Pull request by pKorsholm**
**What** 
- request return revamp

files to review: 
- `src/domain/orders/details/returns/index.js`

*Note: this is dependent on #262*

image

### CODE REVIEW

File: src/domain/orders/details/returns/index.js [hunk 1]
```diff
@@ -136,219 +149,328 @@ const ReturnMenu = ({ order, onReturn, onDismiss, toaster }) => {
...
-                  currency={order.currency_code}
-                  value={refundAmount / 100}
-                  onChange={handleRefundUpdated}
-                />
-              </Box>
-            </Flex>
+                  className="mt-2"
+                  size="small"
+                  currentCurrency={order.currency_code}
+                  readonly
```
Comment (zakariaelas) [inline src/domain/orders/details/returns/index.js:411]: readOnly :smile:

File: src/domain/orders/details/returns/index.js [hunk 2]
```diff
@@ -136,219 +149,328 @@ const ReturnMenu = ({ order, onReturn, onDismiss, toaster }) => {
...
+                        <Table.Row className="last:border-b-0 hover:bg-grey-0">
+                          <Table.Cell colspan={5}>
+                            <div className="w-full flex justify-end">
+                              <Button
+                                variant="ghost"
+                                size="small"
+                                className="border border-grey-20"
+                              >
+                                Select Reason
+                              </Button>
```
Comment (zakariaelas) [inline src/domain/orders/details/returns/index.js:305]: doesn't do anything, that's not part of this PR right?

File: src/domain/orders/details/returns/index.js [hunk 3]
```diff
@@ -136,219 +149,328 @@ const ReturnMenu = ({ order, onReturn, onDismiss, toaster }) => {
...
-            Complete
-          </Button>
+              <span className="ml-3 flex items-center text-grey-90 gap-x-xsmall">
+                Send notifications
+                <InfoTooltip content="Notify customer of created return" />
+              </span>
+            </div>
+            <div className="flex gap-x-xsmall">
+              <Button
+                loading={submitting}
```
Comment (zakariaelas) [inline src/domain/orders/details/returns/index.js:453]: i'd say rm but if you think we should block the user from dismissing the modal, maybe disabling the button would make more sense as opposed to showing a spinner .. what do you think?

**Review by zakariaelas — COMMENTED** [review]
Nice work Philip! LGTM :+1: w. small comments

File: src/domain/orders/details/returns/index.js (hunk 2 above)
Comment (pKorsholm) [reply src/domain/orders/details/returns/index.js]: it is now 😄

File: src/domain/orders/details/returns/index.tsx [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+  useEffect(() => {
+    Medusa.shippingOptions
+      .list({
+        region_id: order.region_id,
+        is_return: true,
+      })
+      .then(({ data }) => {
+        setShippingOptions(data.shipping_options)
+        setShippingLoading(false)
+      })
```
Comment (olivermrbl) [inline src/domain/orders/details/returns/index.tsx:57]: Could we use `useAdminShippingOptions` hook here?

File: src/domain/orders/details/rma-sub-modals/return-reasons.tsx [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+type RMAReturnReasonSubModalProps = {
+  onSubmit: (reason, note) => void
+  reason?: any
+  existingNote?: string
+}
+
+// {
+//   // onSelectReturnReason,
+//   // enableImages,
+// }
```
Comment (olivermrbl) [inline src/domain/orders/details/rma-sub-modals/return-reasons.tsx:18]: rm

File: src/domain/orders/details/rma-sub-modals/return-reasons.tsx [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+  existingNote,
+}) => {
+  const { pop } = useContext(LayeredModalContext)
+  const { isLoading, return_reasons } = useAdminReturnReasons()
+  const [note, setNote] = useState(existingNote || "")
+  const [selectedReason, setSelectedReason] = useState(
+    reason ? { value: reason, label: reason.label } : undefined
+  )
+
+  console.log(return_reasons)
```
Comment (olivermrbl) [inline src/domain/orders/details/rma-sub-modals/return-reasons.tsx:31]: rm

File: src/domain/settings/regions/edit-shipping.tsx [hunk 7]
```diff
@@ -7,49 +7,60 @@ import { useForm } from "react-hook-form"
 import Button from "../../../components/fundamentals/button"
```
Comment (olivermrbl) [inline src/domain/settings/regions/edit-shipping.tsx]: Is this file supposed to be part of this PR? 😊

File: src/domain/orders/details/returns/receive-menu.tsx [hunk 8]
```diff
@@ -0,0 +1,229 @@
+import React, { useState, useEffect } from "react"
```
Comment (olivermrbl) [inline src/domain/orders/details/returns/receive-menu.tsx:1]: Would be great, if we could remove the use of `rebass` in this component

File: src/domain/orders/details/returns/receive-menu.tsx [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+                  >
+                    <Box fontSize={1} px={2}>
+                      Shipping cost
+                    </Box>
+                    <Box px={2} fontSize={1} width={170}>
+                      {(
+                        (returnRequest.shipping_method.price / 100) *
+                        (1 + order.tax_rate / 100)
+                      ).toFixed(2)}{" "}
+                      {order.currency_code.toUpperCase()}
```
Comment (olivermrbl) [inline src/domain/orders/details/returns/receive-menu.tsx:159]: Use one of the prices util functions

File: src/components/molecules/modal/index.tsx [hunk 10]
```diff
@@ -3,13 +3,15 @@ import CrossIcon from "../../fundamentals/icons/cross-icon"
...
-type ModalProps = {
+export type ModalProps = {
   isLargeModal?: boolean
   handleClose: () => void
 }
 
 type ModalChildProps = {
   isLargeModal?: boolean
+  className?: string
+  style?: object
```
Comment (olivermrbl) [inline src/components/molecules/modal/index.tsx:14]: Use `React.CSSProperties` for style

File: src/components/organisms/rma-select-product-table/index.tsx [hunk 11]
```diff
@@ -0,0 +1,262 @@
...
+      (item.quantity - item.returned_quantity === toReturn[item.id].quantity &&
+        change > 0) ||
+      (toReturn[item.id].quantity === 1 && change < 0)
+    ) {
+      return
+    }
+    // const newQuantities = {
+    //   ...quantities,
+    //   [item.id]: quantities[item.id] + change,
+    // }
```
Comment (olivermrbl) [inline src/components/organisms/rma-select-product-table/index.tsx:42]: rm

File: src/components/organisms/rma-select-product-table/index.tsx [hunk 12]
```diff
@@ -0,0 +1,262 @@
...
+  // setQuantities: (quantities: any[]) => void
+}
+
+const RMASelectProductTable: React.FC<RMASelectProductTableProps> = ({
+  order,
+  allItems,
+  toReturn,
+  imagesOnReturns = false,
+  // quantities,
+  // setQuantities,
```
Comment (olivermrbl) [inline src/components/organisms/rma-select-product-table/index.tsx:26]: rm

File: src/components/organisms/rma-select-product-table/index.tsx [hunk 13]
```diff
@@ -0,0 +1,262 @@
...
+import { LayeredModalContext } from "../../molecules/modal/layered-modal"
+import Table from "../../molecules/table"
+
+type RMASelectProductTableProps = {
+  order: any
+  allItems: any[]
+  toReturn: any[]
+  setToReturn: (items: any[]) => void
+  imagesOnReturns: any
+  // setQuantities: (quantities: any[]) => void
```
Comment (olivermrbl) [inline src/components/organisms/rma-select-product-table/index.tsx:17]: rm

File: src/components/organisms/rma-select-product-table/index.tsx [hunk 14]
```diff
@@ -0,0 +1,262 @@
...
+      ...toReturn,
+      [item.id]: {
+        ...toReturn[item.id],
+        quantity: (toReturn[item.id]?.quantity || 0) + change,
+      },
+    }
+
+    setToReturn(newReturns)
+
+    // setQuantities(newQuantities)
```
Comment (olivermrbl) [inline src/components/organisms/rma-select-product-table/index.tsx:54]: rm

File: src/components/organisms/rma-select-product-table/index.tsx [hunk 15]
```diff
@@ -0,0 +1,262 @@
...
+    if (swap_id) {
+      return travFind(order.swaps, swap_id)
+    }
+    if (claim_order_id) {
+      return travFind(order.claims, claim_order_id)
+    }
+    return false
+  }
+
+  console.log(toReturn)
```
Comment (olivermrbl) [inline src/components/organisms/rma-select-product-table/index.tsx:103]: rm

File: src/components/organisms/rma-select-product-table/index.tsx [hunk 16]
```diff
@@ -0,0 +1,262 @@
...
+                      <img
+                        className="h-full w-full object-cover rounded"
+                        src={item.thumbnail}
+                      />
+                    </div>
+                    <div className="inter-small-regular text-grey-50 flex flex-col ml-4">
+                      <span>
+                        <span className="text-grey-90">{item.title}</span> test
+                      </span>
+                      <span>{item.variant.title}</span>
```
Comment (olivermrbl) [inline src/components/organisms/rma-select-product-table/index.tsx:161]: Variant is not necessarily present (it would be in almost all cases, but there is a couple of exceptions like gift cards)
```suggestion
                      <span>{item?.variant?.title || ""}</span>
```

File: src/components/organisms/rma-select-product-table/index.tsx [hunk 17]
```diff
@@ -0,0 +1,262 @@
...
+                      </span>
+                    </div>
+                  ) : (
+                    <span className="text-grey-40">
+                      {item.quantity - item.returned_quantity}
+                    </span>
+                  )}
+                </Table.Cell>
+                <Table.Cell className="text-right">
+                  {(item.refundable / 100).toFixed(2)}
```
Comment (olivermrbl) [inline src/components/organisms/rma-select-product-table/index.tsx:191]: Should account for amounts that should not be divided by 100 like `JPY`

File: src/components/organisms/rma-select-product-table/index.tsx [hunk 18]
```diff
@@ -0,0 +1,262 @@
...
+                </Table.Cell>
+                <Table.Cell className="text-right text-grey-40 pr-1">
+                  {order.currency_code.toUpperCase()}
+                </Table.Cell>
+              </Table.Row>
+              {checked && (
+                <Table.Row className="last:border-b-0 hover:bg-grey-0">
+                  {/* <Table.Cell></Table.Cell> */}
```
Comment (olivermrbl) [inline src/components/organisms/rma-select-product-table/index.tsx:199]: rm

File: src/domain/orders/details/returns/index.tsx [hunk 19]
```diff
@@ -0,0 +1,295 @@
...
+      }
+    }
+
+    if (onReturn) {
+      setSubmitting(true)
+      return onReturn(data)
+        .then(() => onDismiss())
+        .then(() => toaster("Successfully returned order", "success"))
+        .catch((error) => toaster(getErrorMessage(error), "error"))
+        .finally(() => setSubmitting(false))
```
Comment (olivermrbl) [inline src/domain/orders/details/returns/index.tsx:107]: We should consider moving the promise handling to the parent component as part of the `onSuccess` and `onError` functions of the new admin hooks. Let me know what you think?

If we choose to do this, there are a couple of more places to change it.

**Review by olivermrbl — CHANGES_REQUESTED** [review]
Really awesome work @pKorsholm! Love the layered modals 😍 

Added a couple of comments here and there.

File: src/domain/orders/details/returns/index.tsx (hunk 4 above)
Comment (pKorsholm) [reply src/domain/orders/details/returns/index.tsx:57]: not at the moment, this: 
```javascript
const { shipping_options } = useAdminShippingOptions({
    region_id: order.region_id,
    is_return: "true",
  })
```
returns all shipping options for the store

Comment (olivermrbl) [reply src/domain/orders/details/returns/index.tsx:57]: Hmm strange. Do you have any idea why @zakariaelas ?

File: src/domain/orders/details/returns/index.tsx (hunk 19 above)
Comment (pKorsholm) [reply src/domain/orders/details/returns/index.tsx:107]: hmm, it's passed directly from `useMedusa('orders')`, I dont know if we should change it with this pr or fix all of the methods at once?

Comment (olivermrbl) [reply src/domain/orders/details/returns/index.tsx:105]: Since we have the order in this component already, would it be possible for us to use the hook in this component instead?
```
const returnOrder = useAdminRequestReturn(order.id)

return.mutate(.....
```

Comment (olivermrbl) [reply src/domain/orders/details/returns/index.tsx:105]: @pKorsholm

Comment (pKorsholm) [reply src/domain/orders/details/returns/index.tsx]: this does work and i have pushed it, though it doens't update the view now

Comment (pKorsholm) [reply src/domain/orders/details/returns/index.tsx]: @olivermrbl

File: src/domain/orders/details/returns/index.tsx (hunk 4 above)
Comment (kasperkristensen) [reply src/domain/orders/details/returns/index.tsx:58]: @pKorsholm Zak pushed a new canary version of medusa-react the other day that fixes this so try to upgrade the version in your PR should work then :)

**Review by olivermrbl — APPROVED** [review]
LGTM! Just need to resolve a single conflict. Feel free to merge it after @pKorsholm  👍
